lib/spack/spack/cmd/url.py

```diff
@@ -246,6 +257,76 @@ def url_summary(args):
             name_count_dict, version_count_dict)
 
 
+def url_stats(args):
+    stats = {}  # stats about fetchers in packages.
+    total = 0   # total number of versions
+    npkgs = 0   # total number of packages
+
+    def inc(fstype, category, attr=None):
+        """Increment statistics in the stats dict."""
+        if attr:
+            cat_stats = stats.setdefault(fstype, {}).setdefault(category, {})
+            val = cat_stats.setdefault(attr, 0)
+            stats[fstype][category][attr] = val + 1
+        else:
+            val = stats.setdefault(fstype, {}).setdefault(category, 0)
+            stats[fstype][category] = val + 1
```

Comment:
Are you manually designing your own [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict)? Do we have defaultdict backported?

Reply:
For an arbitrarily deep nested defaultdict, you can do something like:
```python
nested_dict = lambda: collections.defaultdict(nested_dict)
stats = nested_dict()
```
But I think yours is fancier, as some levels default to dict and others default to int.

Reply:
Yeah it's fancier (and somewhat uglier) because there are multiple types.  `defaultdict` is actually already imported in `cmd/url.py`.